metrics/metrics.go

```diff
@@ -0,0 +1,37 @@
+// Package metrics provides an extensible framework to instrument your
+// application. All metrics are safe for concurrent use. Considerable design
+// influence has been taken from https://github.com/codahale/metrics and
+// https://prometheus.io.
+package metrics
+
+// Counter is a monotonically-increasing, unsigned, 64-bit integer used to
+// capture the number of times an event has occurred. By tracking the deltas
+// between measurements of a counter over intervals of time, an aggregation
+// layer can derive rates, acceleration, etc.
+type Counter interface {
+	With(Field) Counter
+	Add(delta uint64)
+}
+
+// Gauge captures instantaneous measurements of something using signed, 64-bit
+// integers. The value does not need to be monotonic.
+type Gauge interface {
```

Comment:
Sorry for the insanely long delay, I am just catching up on things now.

The prometheus example actually is a perfect way to show how much easier a callback-based approach would be! Compare what you have there with something like this:

```
// (Obviously I don't feel strongly about the details here... emphasis on the callback aspect)
type GaugeValue interface {
    With(Field) GaugeValue
    SetInt64(value int64)
    SetFloat64(value float64)  // (per the comment about support temperatures, load avg, etc)
}

...

// Then, to define your gauge:
myGauge := newGaugeWithCallback("ops_queued", valCallback func(gv GaugeValue) {
    // (NOTE: the API user can Set() for multiple With() values within this callback)
    gv.SetInt64(len(ops))
})
```

Basically, the value of a metric only matters when it's being sampled by the "monitoring heartbeat", whatever that is in a given monitoring system. Counters are an odd case, because they are actually stateful... but when you want to monitor a gauge, the actual ground truth of the gauge value almost always lives somewhere else in your application, and it makes more sense to just sample that value during the monitoring heartbeat than to Set() and Set() and Set() again every time it changes. Not only is the latter strategy more expensive, but it also leads to incorrect gauge values when a future maintainer doesn't realize that the gauge code needs to be updated every time the thing it's measuring is updated.

Hope this makes sense. Of course a "stateful" gauge is trivial to implement on top of the callback-based approach.